### PR TITLE
Add Allure TestOps support

### PR DESCRIPTION
**Pull request by andrey-brovko**

### CODE REVIEW

File: GherkinSyncTool.Synchronizers.AllureTestOps/AllureTestOpsSynchronizer.cs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                            try
+                            {
+                                addCase = _allureClient.AddTestCase(caseRequest);
+                            }
+                            catch (AllureException e)
+                            {
+                                Log.Error(e, $"The case has not been created: {scenario.Name}");
+                                _context.IsRunSuccessful = false;
+                                continue;
+                            }
```
Comment (vladimir-s-quantori) [inline GherkinSyncTool.Synchronizers.AllureTestOps/AllureTestOpsSynchronizer.cs:90]: Looks like lines 51 - 62 and 78 - 90 are the same. It calls for method extraction

File: GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClient.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                    }
+                )
+                
+            });
+        }
+
+        public IEnumerable<Quantori.AllureTestOpsClient.Model.Content> GetAllTestCases()
+        {
+            var allContent = new List<Quantori.AllureTestOpsClient.Model.Content>();
+            var isLastElementOnThePag = false;
```
Comment (vladimir-s-quantori) [inline GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClient.cs:48]: 'e' is missed in the end of variable name

**Review by alexhurs — APPROVED** [review]

File: GherkinSyncTool.Synchronizers.AllureTestOps/Content/CaseContentBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+            var fullStep = keywordFormatted + stepFormatted;
+
+            if (step.Argument is DocString docString)
+            {
+                fullStep += Environment.NewLine + docString.Content;
+            }
+
+            if (step.Argument is DataTable table)
+            {
```
Comment (vladimir-s-quantori) [inline GherkinSyncTool.Synchronizers.AllureTestOps/Content/CaseContentBuilder.cs:62]: How many steps could be there approximately? Maybe using of StringBuilder will be better here

File: GherkinSyncTool.Synchronizers.AllureTestOps/Content/CaseContentBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        foreach (var step in steps)
+        {
+            var keywordFormatted = $"**{step.Keyword.Trim()}** ";
+            var stepFormatted = step.Text.Replace("<", "***").Replace(">", "***");
+
+            var fullStep = keywordFormatted + stepFormatted;
+
+            if (step.Argument is DocString docString)
+            {
+                fullStep += Environment.NewLine + docString.Content;
```
Comment (vladimir-s-quantori) [inline GherkinSyncTool.Synchronizers.AllureTestOps/Content/CaseContentBuilder.cs:58]: See comment below

File: GherkinSyncTool.Synchronizers.AllureTestOps/Content/CaseContentBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                table.Append($"|{cell.Value}");
+            }
+
+            table.AppendLine();
+        }
+
+        return table.ToString();
+    }
+    
+    // private string ConvertToStringReferences(Scenario scenario, IFeatureFile featureFile)
```
Comment (vladimir-s-quantori) [inline GherkinSyncTool.Synchronizers.AllureTestOps/Content/CaseContentBuilder.cs:102]: Do we need this commented code?

File: Quantori.AllureTestOpsClient/Model/GetTestCasesResponse.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        public int Offset { get; set; }
+        public Sort Sort { get; set; }
+        public int PageSize { get; set; }
+        public bool Unpaged { get; set; }
+        public int PageNumber { get; set; }
+        public bool Paged { get; set; }
+    }
+
+    public class Sort
```
Comment (vladimir-s-quantori) [inline Quantori.AllureTestOpsClient/Model/GetTestCasesResponse.cs:39]: Why not use enum here?

File: Quantori.AllureTestOpsClient/Model/GetTestCasesResponse.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
+﻿using System.Collections.Generic;
+
+namespace Quantori.AllureTestOpsClient.Model
```
Comment (vladimir-s-quantori) [inline Quantori.AllureTestOpsClient/Model/GetTestCasesResponse.cs:3]: In general, we create separate files for separate classes. Any particular reason to combine them in one file?

File: Quantori.AllureTestOpsClient/Model/Link.cs [hunk 8]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Quantori.AllureTestOpsClient.Model
+{
+    public class Link
+    {
+        public string Name { get; set; }
+        public string Url { get; set; }
+        public string Type { get; set; }
```
Comment (vladimir-s-quantori) [inline Quantori.AllureTestOpsClient/Model/Link.cs:7]: What is the "Link type"? If the amount of types is known and limited to a reasonable extent, it makes sense to use enum

File: Quantori.AllureTestOpsClient/Model/Step.cs [hunk 9]
```diff
@@ -0,0 +1,14 @@
+﻿using System.Collections.Generic;
+
+namespace Quantori.AllureTestOpsClient.Model
+{
+    public class Step
+    {
+        public string Name { get; set; }
+        public string Keyword { get; set; }
+        public List<Attachment> Attachments { get; set; }
+        public bool Leaf { get; set; }
```
Comment (vladimir-s-quantori) [inline Quantori.AllureTestOpsClient/Model/Step.cs:10]: What is "leaf"?

File: Quantori.AllureTestOpsClient/Model/TestCaseOverview.cs [hunk 10]
```diff
@@ -0,0 +1,110 @@
+﻿using System.Collections.Generic;
+
+namespace Quantori.AllureTestOpsClient.Model
```
Comment (vladimir-s-quantori) [inline Quantori.AllureTestOpsClient/Model/TestCaseOverview.cs:3]: Again, why all these classes in a single file?

File: Quantori.AllureTestOpsClient/Model/TestCaseOverview.cs [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        public int Id { get; set; }
+        public string Name { get; set; }
+        public ulong CreatedDate { get; set; }
+        public ulong LastModifiedDate { get; set; }
+        public string CreatedBy { get; set; }
+        public string LastModifiedBy { get; set; }
+    }
+
+    public class Mute
+    {
```
Comment (vladimir-s-quantori) [inline Quantori.AllureTestOpsClient/Model/TestCaseOverview.cs:71]: If it is not some Allure term, maybe name it like "MuteReason" or something close

**Review by vladimir-s-quantori — CHANGES_REQUESTED** [review]

File: Quantori.AllureTestOpsClient/Model/GetTestCasesResponse.cs (hunk 6 above)
Comment (andrey-brovko) [reply Quantori.AllureTestOpsClient/Model/GetTestCasesResponse.cs:39]: That is a server API contract that will be serialized to Json.

File: Quantori.AllureTestOpsClient/Model/GetTestCasesResponse.cs (hunk 7 above)
Comment (andrey-brovko) [reply Quantori.AllureTestOpsClient/Model/GetTestCasesResponse.cs:3]: It was auto-generated by swagger -> class converted. I am going to move the inner classes to separate files later.

File: Quantori.AllureTestOpsClient/Model/Link.cs (hunk 8 above)
Comment (andrey-brovko) [reply Quantori.AllureTestOpsClient/Model/Link.cs:7]: Let's discuss it offline. I am not sure that we are on the same page.

File: GherkinSyncTool.Synchronizers.AllureTestOps/Content/CaseContentBuilder.cs (hunk 5 above)
Comment (andrey-brovko) [reply GherkinSyncTool.Synchronizers.AllureTestOps/Content/CaseContentBuilder.cs]: Nope. Removed.

File: GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClient.cs (hunk 2 above)
Comment (andrey-brovko) [reply GherkinSyncTool.Synchronizers.AllureTestOps/Client/AllureClient.cs]: Fixed

File: GherkinSyncTool.Synchronizers.AllureTestOps/AllureTestOpsSynchronizer.cs [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+                    var caseRequest = _caseContentBuilder.BuildCaseRequest(scenario, featureFile);
+
+                    // Create test case for feature file which is getting synced for the first time, so no tag id present.  
+                    if (tagId is null)
+                    {
+                        TestCase addCase;
+                        try
+                        {
+                            addCase = _allureClient.AddTestCase(caseRequest);
```
[the remark this answers is not in the document]
Comment (andrey-brovko) [reply GherkinSyncTool.Synchronizers.AllureTestOps/AllureTestOpsSynchronizer.cs]: Fixed

File: GherkinSyncTool.Synchronizers.AllureTestOps/AllureTestOpsSynchronizer.cs [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        public void Sync(List<IFeatureFile> featureFiles)
+        {
+            var stopwatch = Stopwatch.StartNew();
+            Log.Info("# Start synchronization with Allure TestOps");
+
+            var allureTestCases = _allureClient.GetAllTestCases().ToList();
+            
+            foreach (var featureFile in featureFiles)
+            {
+                var insertedTagIds = 0;
```
[the remark this answers is not in the document]
Comment (andrey-brovko) [reply GherkinSyncTool.Synchronizers.AllureTestOps/AllureTestOpsSynchronizer.cs]: Done. Actually, it is outside of the loop which uses the variable.

File: GherkinSyncTool.Synchronizers.AllureTestOps/AllureTestOpsSynchronizer.cs [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+                
+                foreach (var scenario in featureFile.Document.Feature.Children.OfType< Scenario>())
+                {
```
[the remark this answers is not in the document]
Comment (andrey-brovko) [reply GherkinSyncTool.Synchronizers.AllureTestOps/AllureTestOpsSynchronizer.cs:44]: Fixed.

File: GherkinSyncTool.Synchronizers.AllureTestOps/AllureTestOpsSynchronizer.cs (hunk 1 above)
Comment (andrey-brovko) [reply GherkinSyncTool.Synchronizers.AllureTestOps/AllureTestOpsSynchronizer.cs]: Done

**Review by vladimir-s-quantori — APPROVED** [review]

**Pull request merged**